dataset/groupby.cpp

```diff
@@ -323,16 +323,31 @@ template <class T> struct MakeBinGroups {
   }
 };
 
+template <class T, class U>
+GroupBy<T> callDType(const U &array, const VariableConstView &key,
+                     const VariableConstView &bins) {
```

Comment:
Good idea extracting the comment code, but a couple improvement suggestions:
- `callDType` is a generic helper, here a more specific name would be better. `call_groupby` or something like that maybe?
- You can avoid one of the template arguments, since basically `U` defines `T`. Basically: replace `T` by `typename U::value_type`. I notice that `value_type` is actually missing from `DataArrayConstView` and `DatasetConstView` (probably it was never needed before). I can be added, similar to https://github.com/scipp/scipp/blob/c97e165cf9cbe96a7c7a983def350148cb45a5de/variable/include/scipp/variable/variable.h#L233 (replacing `Variable` by `DataArray` and `Dataset`, respectively).

Finally, when you *call* `call_groupby` below, you can omit the template arguments, since `U` will be deduced.

Reply:
> since U will be deduced.

Wow. How? I thought I knew how to use templates and then suddenly gems like this :)
Thanks, Simon. I'm learning a lot.

Reply:
Magic :)
Since C++17 it amazingly also works for classes, see https://en.cppreference.com/w/cpp/language/class_template_argument_deduction.